src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs

```diff
@@ -164,34 +164,42 @@ protected override Expression VisitScoped(ScopedCallSite callSite, ParameterExpr
                 callSite.Key,
                 typeof(object));
 
-            var resolvedExpression = Expression.Variable(typeof(object), "resolved");
+            var resolvedVariable = Expression.Variable(typeof(object), "resolved");
 
             var resolvedServices = GetResolvedServices(provider);
 
             var tryGetValueExpression = Expression.Call(
                 resolvedServices,
                 TryGetValueMethodInfo,
                 keyExpression,
-                resolvedExpression);
+                resolvedVariable);
 
             var assignExpression = Expression.Assign(
-                resolvedExpression, VisitCallSite(callSite.ServiceCallSite, provider));
+                resolvedVariable, VisitCallSite(callSite.ServiceCallSite, provider));
 
             var addValueExpression = Expression.Call(
                 resolvedServices,
                 AddMethodInfo,
                 keyExpression,
-                resolvedExpression);
+                resolvedVariable);
+
+            Expression captureDisposableExpression = Expression.Invoke(
```

Comment:
We never tested the actual performance in terms of op/s, just memory allocations. If you look closely at your [original PR](https://github.com/aspnet/DependencyInjection/pull/291) you'll notice that the memory allocations measured are mostly related ServiceProvider.CreateScope()/ctor, and there are no measurements related to the actual capturing disposables of any kind.

If you're really concerned memory you could store some sort of ordering metadata with the captured services (a la Autofac) or pool the Dictionary/List like @benaadams [suggested in the original PR](https://github.com/aspnet/DependencyInjection/pull/291#issuecomment-140784285). Both sound kinda complicated and might even hurt perf in some scenarios.

It might be a good idea to add a BenchmarkDotNet scenario that covers capturing disposables.